cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramLifecycleService.java

```diff
@@ -525,10 +528,30 @@ public RunId runInternal(ProgramId programId, Map<String, String> userArgs, Map<
 
     checkCapability(programDescriptor);
     synchronized (this) {
-      if (maxConcurrentRuns > 0 && maxConcurrentRuns <= store.countActiveRuns(maxConcurrentRuns)) {
-        ConflictException e = new ConflictException(
-          String.format("Program %s cannot start because the maximum of %d concurrent runs is exceeded",
-                        programId, maxConcurrentRuns));
+      Map<ProgramRunStatus, Integer> activeRunCounts = store.countActiveRuns(maxConcurrentRunning);
```

Comment:
We disabled the use of countActiveRuns (by always setting max concurrent runs to -1) because it was extremely slow, which also runs inside a sync block. It affect the performance of concurrent launch a lot.

Reply:
Changed the logic to read from in-memory DS.